R/pmcmc.R

```diff
@@ -0,0 +1,428 @@
+    tmp <- rnorm(n = n_pars, mean = pars, sd = pars_sd)
+    # discretise if necessary
+    tmp[pars_discrete] <- round(tmp[pars_discrete])
+    # reflect proposal if it exceeds upper or lower parameter boundary
+    tmp <- reflect_proposal(x = tmp, 
+                            floor = pars_min, 
+                            cap = pars_max)
+    tmp
+  }
+
+
+  # start progress bar  
+  pb <- txtProgressBar(min = 0, max = n_mcmc, style = 3)
+  
+  # main pmcmc loop
+  for(iter in seq_len(n_mcmc)) {
+    
+    setTxtProgressBar(pb, iter)
+    
+    # propose new parameters
+    prop_pars <- propose_jump(curr_pars)
+    
+    ## calculate proposed prior / lhood / posterior 
+    prop_lprior <- calc_lprior(pars = prop_pars)
+    p_filter_est <- calc_ll(pars = curr_pars)
+    prop_ll <- p_filter_est$log_likelihood
+    prop_ss <- p_filter_est$sample_state
+    prop_lpost <- prop_lprior + prop_ll
+    
+    
+    # calculate probability of acceptance
+    accept_prob <- exp(prop_lpost - curr_lpost)
+
+    
+    if(runif(1) < accept_prob) { # MH step
+      # update parameters and calculated likelihoods
+      curr_pars <- prop_pars
+      curr_lprior <- prop_lprior
+      curr_ll <- prop_ll
+      curr_lpost <- prop_lpost
+      curr_ss <- prop_ss
+    }
+    
+    # record results
+    res[iter, ] <- c(curr_pars, 
+                     'log_prior' = curr_lprior, 
+                     'log_likelihood' = curr_ll, 
+                     'log_posterior' = curr_lpost) 
+    states[iter, ] <- curr_ss
+
+  }
+  # end progress bar
+  close(pb)
+  
+  
+  # include starting conditions in output
+  res <- rbind(out_init, res, deparse.level = 0)
+  res <- as.data.frame(res)
+  
+  tmp <- coda::as.mcmc(res)
+  rejection_rate <- coda::rejectionRate(tmp)
+  ess <- coda::effectiveSize(tmp)
+  
+  res$start_date <- as.Date(data$date[1]) - res$start_date
+  
+  states <- rbind(ss_init, states, deparse.level = 0)
+
+  
+  
+ list('inputs' = inputs, 
+      'results' = as.data.frame(res),
+      'states' = states, 
+      'acceptance_rate' = 1-rejection_rate, 
+      "ess" = ess
+      )
+
+}
+
+
+calc_loglikelihood <- function(pars, data, model_params,  steps_per_day, pars_obs, n_particles) {
+  
+  start_date <- as.Date(data$date[1]) - pars[['start_date']]
+  model_params$beta_y <-  pars[['beta']]
+  
+  # create sir_covid model
+  mod <- sircovid(params = model_params)
+
+  d <- particle_filter_data(data, 
+                            start_date = start_date, 
+                            steps_per_day = steps_per_day)
+  compare <- compare_icu(mod, pars_obs, d)
+  X <- particle_filter(data = d, 
+                       model = mod, 
+                       compare = compare, 
+                       n_particles = n_particles)
+  # This is currently a placeholder for when the particle filter function in the sircovid package
+  # function is changed to allow output of a sample state 
+  X$sample_state <- rep(1, 238)
```

Comment:
I think your change to allow this is now merged, can this be updated to use it?

Reply:
change made and pushed